File: backend-rust/src/core/approx_engine.rs
use serde_json::Value;

use crate::core::executor::build_sample_query;
use crate::core::parser::parse_analytical_query;
use crate::core::runtime_sampling::{
    run_runtime_sampling,
    mode_configs,
};

pub fn rewrite_agg_query(
    query: &str,
    source: &str,
    mode: &str,
) -> Result<String, String> {
    let source_key = source.trim().to_lowercase();

    let configs = mode_configs();

    let mode_key = if configs.contains_key(mode) {
        mode
    } else {
        "balanced"
    };

    let parsed = parse_analytical_query(query)?;

    let first_fraction = configs
        .get(mode_key)
        .ok_or("Mode config not found")?
        .progression[0];

    Ok(
        build_sample_query(
            &parsed,
            &source_key,
            first_fraction,
        )
    )
}

pub fn run_approx(
    query: &str,
    source: &str,
    mode: &str,
    accuracy_target: Option<f64>,
) -> Result<Value, String> {
    let source_key = source.trim().to_lowercase();

    let parsed =
        parse_analytical_query(query)?;

    let result = run_runtime_sampling(
        &parsed,
        &source_key,
        mode,
        accuracy_target,
    );

    serde_json::to_value(result)
        .map_err(|e| e.to_string())
}

File: backend/core/approx_engine.py
from __future__ import annotations

from typing import Any, Callable

from backend.core.executor import build_sample_query
from backend.core.parser import parse_analytical_query
from backend.core.runtime_sampling import MODE_CONFIGS, run_runtime_sampling


def _rewrite_agg_query(query: str, source: str, mode: str = "balanced") -> str:
    source_key = source.lower().strip()
    mode_key = mode if mode in MODE_CONFIGS else "balanced"
    parsed = parse_analytical_query(query)
    first_fraction = MODE_CONFIGS[mode_key]["progression"][0]
    return build_sample_query(parsed, source_key, first_fraction)


def run_approx(
    query: str,
    source: str = "duckdb",
    mode: str = "balanced",
    accuracy_target: float | None = None,
    progress_callback: Callable[[dict[str, Any]], None] | None = None,
    ci_multiplicity_correction: bool = True,
    ci_anytime_valid: bool = True,
    ci_coverage_level: float = 0.95,
) -> dict[str, Any]:
    source_key = source.lower().strip()
    parsed = parse_analytical_query(query)
    return run_runtime_sampling(
        parsed,
        source_key,
        mode,
        accuracy_target=accuracy_target,
        progress_callback=progress_callback,
        ci_multiplicity_correction=ci_multiplicity_correction,
        ci_anytime_valid=ci_anytime_valid,
        ci_coverage_level=ci_coverage_level,
    )
